File: HelloMetal-OC/HelloMetal-OC/Renderer.m
//
//  Renderer.m
//  HelloMetal-OC
//
//  Created by iann on 2019/7/23.
//  Copyright © 2019 luciglobal. All rights reserved.
//
#import <Cocoa/Cocoa.h>
#import "Renderer.h"
#import <Metal/Metal.h>
#import <MetalKit/MetalKit.h>
#import "ShaderTypes.h"

@implementation Renderer {
    id<MTLBuffer> vertexBuffer;
    id <MTLRenderPipelineState> pipelineState;
    id<MTLTexture> texture;
}
- (instancetype) init:(MTKView *) mtkView
{
    device = mtkView.device;
    queue = device.newCommandQueue;

    
    static const MyVertex vertexArrayData[] = {
        {{0.5,-0.25,0,1},{0.75,0.25}},
        {{-0.5,-0.25,0,1},{0.25,0.25}},
        {{0,0.5,0,1},{0,0.75}},
    };
    vertexBuffer = [device newBufferWithBytes:vertexArrayData length:sizeof(vertexArrayData) options:0];
    
    
    
    MTLRenderPipelineDescriptor *des = [MTLRenderPipelineDescriptor new];
    id<MTLLibrary> library = [device newDefaultLibrary];
    des.vertexFunction = [library newFunctionWithName:@"myVertexShader"];
    des.fragmentFunction = [library newFunctionWithName:@"myFragmentShader"];
    des.colorAttachments[0].pixelFormat = mtkView.colorPixelFormat;
    NSError *error;
    pipelineState = [device newRenderPipelineStateWithDescriptor:des error:&error];
    
    [self loadTexture];
    return self;
}

- (void) loadTexture
{
    
    NSError *error;
//    NSImage *image = [NSImage imageNamed:@"abc"];
//    MTLTextureDescriptor *textureDescriptor = [[MTLTextureDescriptor alloc] init];
//    textureDescriptor.pixelFormat = MTLPixelFormatRGBA8Unorm;
//    textureDescriptor.width = image.size.width;
//    textureDescriptor.height = image.size.height;
//    texture = [device newTextureWithDescriptor:textureDescriptor];
//    //self.texture = [self.mtkView.device newTextureWithDescriptor:textureDescriptor]; // 创建纹理
    
    MTKTextureLoader* textureLoader = [[MTKTextureLoader alloc] initWithDevice:device];
    
    NSDictionary *textureLoaderOptions =
    @{
      MTKTextureLoaderOptionTextureUsage       : @(MTLTextureUsageShaderRead),
      MTKTextureLoaderOptionTextureStorageMode : @(MTLStorageModePrivate)
      };
    
    texture = [textureLoader newTextureWithName:@"abc"
                                      scaleFactor:1.0
                                           bundle:nil
                                          options:textureLoaderOptions
                                            error:&error];
    int a = 0;
}

- (void)drawInMTKView:(nonnull MTKView *)view
{
    id<MTLCommandBuffer> buffer = queue.commandBuffer;
    MTLRenderPassDescriptor *passDesc = view.currentRenderPassDescriptor;
    passDesc.colorAttachments[0].clearColor = MTLClearColorMake(1, 0, 0, 1);
    
    
    id <MTLRenderCommandEncoder> encorder = [buffer renderCommandEncoderWithDescriptor: passDesc];
    [encorder setCullMode:MTLCullModeNone];
    [encorder setFrontFacingWinding:MTLWindingCounterClockwise];
    [encorder setRenderPipelineState: pipelineState];
    [encorder setVertexBuffer:vertexBuffer offset:0 atIndex:0];
    [encorder setFragmentTexture:texture atIndex:0];
    [encorder drawPrimitives:MTLPrimitiveTypeTriangle vertexStart:0 vertexCount:3];
    [encorder endEncoding];
    [buffer presentDrawable:view.currentDrawable];
    [buffer commit];
    
}

- (void)mtkView:(nonnull MTKView *)view drawableSizeWillChange:(CGSize)size
{
    id<MTLCommandBuffer> commandBuffer = [queue commandBuffer];
    
    
}




@end

File: HelloMetal-Swift/HelloMetal-Swift/Renderer.swift
//
//  Renderer.swift
//  HelloMetal-Swift
//
//  Created by iann on 2019/7/22.
//  Copyright © 2019 luciglobal. All rights reserved.
//

import Foundation
import Metal
import MetalKit

class Renderer : NSObject, MTKViewDelegate {
    
    let device: MTLDevice
    let commandQueue: MTLCommandQueue
    let pipelineState: MTLRenderPipelineState
    let vertexBuffer: MTLBuffer
    
    // This is the initializer for the Renderer class.
    // We will need access to the mtkView later, so we add it as a parameter here.
    init?(mtkView: MTKView) {
        device = mtkView.device!
        commandQueue = device.makeCommandQueue()!
        
        do {
            pipelineState = try Renderer.buildRenderPipelineWith(device: device, metalKitView: mtkView)
        }
        catch{
            print("Unable to compile render pipeline state: \(error)")
            return nil
        }
        
        let vertices = [Vertex(color: [1,0,0,1], pos: [-1,-1]),
                        Vertex(color: [1,1,0,1], pos: [0,1]),
                        Vertex(color: [1,1,1,1], pos: [1,-1])]
        
        
        
        vertexBuffer = device.makeBuffer(bytes: vertices, length: vertices.count * MemoryLayout<Vertex>.stride, options: [])!
        
        
        
    }
    
    // Create our custom rendering pipeline, which loads shaders using 'device', and outputs to the format of 'metalKitView'
    class func buildRenderPipelineWith(device :MTLDevice, metalKitView: MTKView) throws -> MTLRenderPipelineState {
        // Create a new pipeline descriptor
        let pipelineDescriptor = MTLRenderPipelineDescriptor()
        
        let library = device.makeDefaultLibrary()
        pipelineDescriptor.vertexFunction = library?.makeFunction(name: "vertexShader")
        pipelineDescriptor.fragmentFunction = library?.makeFunction(name: "fragmentShader")
        
        pipelineDescriptor.colorAttachments[0].pixelFormat = metalKitView.colorPixelFormat
        return try device.makeRenderPipelineState(descriptor: pipelineDescriptor)
    }
    
    // mtkView will automatically call this function
    // whennever it wants new content to be rendered.
    func draw(in view: MTKView) {
        
        guard let commandBuffer = commandQueue.makeCommandBuffer() else { return }
        
        guard let renderPassDescriptor = view.currentRenderPassDescriptor else { return }
        
        renderPassDescriptor.colorAttachments[0].clearColor = MTLClearColorMake(1,0,0,1)
        
        guard let renderEncoder = commandBuffer.makeRenderCommandEncoder(descriptor: renderPassDescriptor) else { return }
        renderEncoder.setRenderPipelineState(pipelineState)
        renderEncoder.setVertexBuffer(vertexBuffer, offset: 0, index: 0)
        renderEncoder.drawPrimitives(type: .triangle, vertexStart: 0, vertexCount: 3)
        renderEncoder.endEncoding()
        

        
        commandBuffer.present(view.currentDrawable!)
        
        commandBuffer.commit()
        
    }
    
    // mtkView will automatically call this funcion
    // whenever the size of view changeds (such as resizing the window).
    func mtkView(_ view: MTKView, drawableSizeWillChange size: CGSize) {
        
    }
}
